File: solutions/go/protein-translation/1/protein_translation.go
//This package translates RNA seuences into proteins
package protein

import ("fmt"
"errors")

const ErrStop=1
const ErrInvalidBase=2

// The FromRNA breaks an RNA String into multiple Codons
func FromRNA(rna string) ([]string, error) {
 var result []string
 for i:= 0; i < len(rna); i =i+3 {
	fmt.Println(rna[3*i:3*i+3])
	result[i/3] = rna[3*i:3*i+3]
 }
 return result, nil
}

// The FromCodon translate into protein
func FromCodon(codon string) (string, error) {
	var ctop = map[string]string {
		"AUG": "Methionine", "UUU":"Phenylalanine", "UUC": "Phenylalanine",
		"UUA":"Leucine", "UUG":"Leucine", "UCG": "Serine",
		"UAU":"Tyrosine","UAC":"Tyrosine","UGU":"Cysteine",
		"UGG":"Tryptophan","UAA":"","UAG":"","UGA":"" }
	protein, ok := ctop[codon]
	if !ok {
		return "", errors.ErrInvalidBase
	}
	if ctop[codon] == "" {
		return "", errors.ErrStop
	}
	return protein, nil
}

File: solutions/fortran/protein-translation/1/protein_translation.f90
module protein_translation
  implicit none

contains

  function proteins(rna) result(names)
    character(len=*), intent(in) :: rna
    character(len=13), allocatable :: names(:)
    character(len=13), dimension(20) :: tnames
    integer :: i,j
    !print *, "rna=", rna
    tnames(:) = ""
    j = 0
    do i = 1,len(trim(rna)),3
        j = j + 1
        select case(rna(i:i+2))
           case("AUG")
                  tnames(j:j) = "Methionine" 
           case("UUU","UUC")
                  tnames(j:j) = "Phenylalanine"
           case("UUA", "UUG")
                 tnames(j:j) =  "Leucine"
           case("UCU", "UCC", "UCA", "UCG")
                 tnames(j:j) =   "Serine"
            case("UAU", "UAC")
                 tnames(j:j) =   "Tyrosine"
            case("UGU", "UGC")
                 tnames(j:j) =   "Cysteine"
            case("UGG")
                 tnames(j:j) =   "Tryptophan"
            case("UAA", "UAG", "UGA")
                    j = j - 1
                    exit 
        end select

    end do    
    !print *, "j=", j
    !print *, "tnames=", tnames
    allocate(names(j))
    if (j>0) names(1:j) = tnames(1:j)
  end function proteins

end module protein_translation

